### ISSUE TITLE
Possible bug in cosine similarity dimension in WMReward/utils.py

### ISSUE THREAD

**Issue by messimm** [facebookresearch/WMReward#11]
In WMReward/utils.py (around line 850), the loss is computed as:

loss = 1 - F.cosine_similarity(z, h[0], dim=1).mean()

I think dim=1 may be incorrect here, and that this should probably be dim=-1.

My reasoning is that z and h[0] appear to have shape [B, N, D], where the last dimension is the feature/embedding dimension. In that case, cosine similarity should be computed over D, not over the token dimension N.

So I would expect:

loss = 1 - F.cosine_similarity(z, h[0], dim=-1).mean()

Could you confirm whether dim=1 is intentional, or if this is a bug?

**Comment by felifri** [comment]
Hi Messimm,

yes, that is correct. That is a typo in the code and we corrected it. The experiments in our internal code base ran with dim=-1. Most of our methods use only 2D Tensors where dim=1 and dim=-1 have the same effect. 

Best,
Felix
